_posts/got-the-time.md

```diff
@@ -0,0 +1,244 @@
+- Invoice paid
+
+We might find ourselves in a situation where the timeout is scheduled, but there is no saga to dispatch the message to because it's already completed. NServiceBus handles the scenario for us; when the timeout expires, the message is automatically discarded if the saga doesn't exist. The assumption is that an existing saga can only schedule a timeout for itself (there is no way to schedule timeouts for a different saga). If the saga doesn't exist anymore, it's fair to assume that it's already completed. The timeout doesn't make sense anymore.
+
+There are other scenarios in which message ordering is essential from the business perspective.
+
+Imagine the case in which customers are paying with a credit card at the time of the order. The generated invoice is immediately paid. The `InvoicePaid` event may arrive before the `InvoiceIssued`. Remember that physical message ordering is extremely hard, if ever possible, in distributed systems. With the above saga, the described scenario surfaces a problem. The saga is designed to be started only by the `InvoceIssued` message. If `InvocePaid` is dispatched first, NServiceBus cannot find an existing saga; the only viable assumption is that the saga is already completed, and thus the message can be discarded. Later NServiceBus processes the `InvoiceIssued` message and creates the saga. It is a data/message loss scenario.
+
+Suppose we find ourselves in a case like the one described above, which is true for most scenarios. In that case, we want to make sure that NServiceBus can start the saga regardless of messages arrival order:
+
+```csharp
+class OverdueInvoicePolicy :
+   Saga<OverdueInvoicePolicy.OverdueInvoiceData>,
+   IAmStartedByMessages<InvoiceIssued>,
+   IAmStartedByMessages<InvoicePaid>,
+   IHandleTimeouts<CheckPayment>
+{
+   public async Task Handle(InvoiceIssued message, IMessageHandlerContext context)
+   {
+      //...
+   }
+
+   public async Task Handle(InvoicePaid message, IMessageHandlerContext context)
+   {
+      //...
+   }
+
+}
+```
+
+All the messages that can start the saga are marked as `IAmStartedByMessages<>`. The next step is to explain to NServiceBus how to correlated both messages:
+
+```csharp
+protected override void ConfigureHowToFindSaga(SagaPropertyMapper<OverdueInvoiceData> mapper)
+{
+   mapper.MapSaga(d => d.InvoiceNumber).ToMessage<InvoiceIssued>(m => m.InvoiceNumber);
+   mapper.MapSaga(d => d.InvoiceNumber).ToMessage<InvoicePaid>(m => m.InvoiceNumber);
+}
+```
+
+Finally, we need a way for the saga to determine when it received all the messages. We can easily do that by storing that information in the saga data type:
+
+```csharp
+class OverdueInvoiceData : ContainSagaData
+{
+   public int InvoiceNumber { get; set; }
+   public bool InvoiceIssuedReceived { get;set; }
+   public bool InvoicePaidReceived { get;set; }
+}
+```
+
+And use the two new flags in the code that handles incoming messages. 
```

Comment:
I know it's simple but I feel a quick code sample showing these flags being set in the `Handle` methods is warranted.